### DISCUSSION TITLE
Dynamic cells with map-cells option

### DISCUSSION THREAD

**Discussion by KaliszS** [PgBiel/typst-tablex#122] [Q&A] [answered]
Hello, is that possible to make cells template dynamic if I use map-cells option? I've got code like that:

```typst
tablex(
    columns: cols_number,
    column-gutter: 10pt,
    auto-vlines: false,
    auto-hlines: false,
    align: center + horizon,
    map-cells: cell => {
      lang(
        langs.at(
          cell.x + cell.y * cols_number
        ),
        level
      )
    },
    
    [], [], [], [], [], [], [], [], [], []
  )
``` 

But since it's part of the function called multiple times for different number of cells, I would like to cut the last line: `[], [], [], [], [], [], [], [], [], []`. Thanks in advance for your help :)

I'm using tablex v 0.0.8 btw.

**Comment by PgBiel** [comment]
You can use `cellx(x: cols_number - 1)[]` to the same effect.

Or in general `cellx(y: 0)[]` to ensure at least one row is present, `cellx(y: 1)[]` to ensure that at least two rows are present, and so on.
Otherwise map-cells cannot operate over unspecified cells.

**Reply by KaliszS** [reply]
It doesn't work because it only displays one row/column. I would like to display all of them, assuming that there may be a different number of them.

Perhaps I have over-complicated this and it can be done simpler than with map-cells?

**Reply by PgBiel** [reply, marked as answer]
Do you mean you just want to display all elements of `langs` as cells in the table? If so, you can use spreading of arrays into arguments[^1], combined with `map`[^2] to transform the array:

```typ
#tablex(
  // ... options and stuff ...
  ..langs.map(x => lang(x, level))
)
```

[^1]: https://typst.app/docs/reference/foundations/arguments/
[^2]: https://typst.app/docs/reference/foundations/array/#definitions-map

**Reply by KaliszS** [reply]
That's exactly what I meant, which is just as I thought, I overcomplicated. Thank you!
